### ISSUE TITLE
netopeer2-server segfaults via sysrepo in __pthread_mutex_unlock_full

### ISSUE THREAD

**Issue by jktjkt** [sysrepo/sysrepo#2312]
Hi there, we're getting this segfault when connecting to netopeer2 via [`netconf-cli`](https://github.com/CESNET/netconf-cli). The crash happens very soon, after the YANG schemas have been requested, upon sending the following operation:
```xml
<rpc xmlns="urn:ietf:params:xml:ns:netconf:base:1.0" message-id="4">
  <get xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
    <filter type="xpath" xmlns:yanglib="urn:ietf:params:xml:ns:yang:ietf-yang-library" select="/yanglib:*"/>
  </get>
</rpc>
```

Here's the backtrace:
```gdb
Core was generated by `/usr/bin/netopeer2-server'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0xb6e4d4c0 in __pthread_mutex_unlock_full () from /lib/libpthread.so.0
[Current thread is 1 (Thread 0xb66f1450 (LWP 488))]
(gdb) t a a bt

Thread 7 (Thread 0xb44dc450 (LWP 492)):
#0  0xb6bf3cb4 in ?? () from /lib/libc.so.6
#1  0xb6bf3cfc in clock_nanosleep () from /lib/libc.so.6
#2  0xb6bf9b50 in nanosleep () from /lib/libc.so.6
#3  0x00019424 in np_sleep (ms=200) at /home/ci/build/build/netopeer2-custom/src/common.c:50
#4  0x00018150 in worker_thread (arg=0xfe8ef8) at /home/ci/build/build/netopeer2-custom/src/main.c:998
#5  0xb6e49144 in start_thread () from /lib/libpthread.so.0
#6  0xb6c2e328 in ?? () from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 6 (Thread 0xb6f4c220 (LWP 485)):
#0  0xb6bf3cb4 in ?? () from /lib/libc.so.6
#1  0xb6bf3cfc in clock_nanosleep () from /lib/libc.so.6
#2  0xb6bf9b50 in nanosleep () from /lib/libc.so.6
#3  0x00019424 in np_sleep (ms=200) at /home/ci/build/build/netopeer2-custom/src/common.c:50
#4  0x00018150 in worker_thread (arg=0xfe93f8) at /home/ci/build/build/netopeer2-custom/src/main.c:998
#5  0x00019214 in main (argc=1, argv=0xbef44e44) at /home/ci/build/build/netopeer2-custom/src/main.c:1337

Thread 5 (Thread 0xb3cdb450 (LWP 493)):
#0  0xb6e4dee4 in __pthread_mutex_cond_lock_full () from /lib/libpthread.so.0
#1  0xb6e51278 in pthread_cond_timedwait@@GLIBC_2.4 () from /lib/libpthread.so.0
#2  0xb6d03848 in sr_shmsub_notify_wait_wr (sub_shm=sub_shm@entry=0xb46dc000, shm_struct_size=shm_struct_size@entry=176, expected_ev=expected_ev@entry=SR_SUB_EV_ERROR, timeout_ms=timeout_ms@entry=1, cid=75, cb_err_info=0xb3cdac14, 
    cb_err_info@entry=0xb671fa20) at /home/ci/build/build/sysrepo-custom/src/shm_sub.c:248
#3  0xb6d066e0 in sr_shmsub_rpc_notify (conn=0xfd2a58, shm_rpc=0xb671f9f8, shm_rpc@entry=0xb6ccfcbc, op_path=0xb3100da0 "/ietf-netconf:get", input=input@entry=0xb3100ea0, sid=..., timeout_ms=timeout_ms@entry=2000, request_id=0xb3cdac24, 
    request_id@entry=0xb3cdac1c, output=output@entry=0xb3cdac90, cb_err_info=0xb3cdac14, cb_err_info@entry=0xb3cdac0c) at /home/ci/build/build/sysrepo-custom/src/shm_sub.c:1620
#4  0xb6cdb62c in sr_rpc_send_tree (session=0x10b2eb0, input=0xb3100ea0, timeout_ms=2000, output=0xb3cdac90) at /home/ci/build/build/sysrepo-custom/src/sysrepo.c:3882
#5  0x00015c38 in np2srv_rpc_cb (rpc=0xb3100ea0, ncs=0x10379c8) at /home/ci/build/build/netopeer2-custom/src/main.c:444
#6  0xb6eee3b8 in nc_server_send_reply_io (rpc=0xb3100e70, io_timeout=10, session=0x10379c8) at /home/ci/build/build/libnetconf2-custom/src/session_server.c:1305
#7  nc_ps_poll (ps=<optimized out>, timeout=10, session=0x0) at /home/ci/build/build/libnetconf2-custom/src/session_server.c:1656
#8  0x00018124 in worker_thread (arg=0xfe9408) at /home/ci/build/build/netopeer2-custom/src/main.c:994
#9  0xb6e49144 in start_thread () from /lib/libpthread.so.0
#10 0xb6c2e328 in ?? () from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 4 (Thread 0xb56de450 (LWP 490)):
#0  0xb6bf3cb4 in ?? () from /lib/libc.so.6
#1  0xb6bf3cfc in clock_nanosleep () from /lib/libc.so.6
#2  0xb6bf9b50 in nanosleep () from /lib/libc.so.6
#3  0x00019424 in np_sleep (ms=200) at /home/ci/build/build/netopeer2-custom/src/common.c:50
#4  0x00018150 in worker_thread (arg=0xfed780) at /home/ci/build/build/netopeer2-custom/src/main.c:998
#5  0xb6e49144 in start_thread () from /lib/libpthread.so.0
#6  0xb6c2e328 in ?? () from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 3 (Thread 0xb4edd450 (LWP 491)):
#0  0xb6bf3cb8 in ?? () from /lib/libc.so.6
#1  0xb6bf3cfc in clock_nanosleep () from /lib/libc.so.6
#2  0xb6bf9b50 in nanosleep () from /lib/libc.so.6
#3  0xb6c28630 in usleep () from /lib/libc.so.6
#4  0xb6eedb04 in nc_ps_poll (ps=0xfe8e28, timeout=10, session=0x0) at /home/ci/build/build/libnetconf2-custom/src/session_server.c:1610
#5  0x00018124 in worker_thread (arg=0xff0068) at /home/ci/build/build/netopeer2-custom/src/main.c:994
#6  0xb6e49144 in start_thread () from /lib/libpthread.so.0
#7  0xb6c2e328 in ?? () from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 2 (Thread 0xb5eea450 (LWP 489)):
#0  0xb6c27e90 in select () from /lib/libc.so.6
#1  0xb6d09174 in sr_shmsub_listen_thread (arg=0xfee938) at /home/ci/build/build/sysrepo-custom/src/shm_sub.c:3333
#2  0xb6e49144 in start_thread () from /lib/libpthread.so.0
#3  0xb6c2e328 in ?? () from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 1 (Thread 0xb66f1450 (LWP 488)):
#0  0xb6e4d4c0 in __pthread_mutex_unlock_full () from /lib/libpthread.so.0
#1  0xb6ce238c in sr_rwunlock (rwlock=rwlock@entry=0xb46d8000, timeout_ms=timeout_ms@entry=0, mode=mode@entry=SR_LOCK_WRITE, cid=75, func=0xb6d0f71c <__func__.3> "sr_shmsub_rpc_listen_process_rpc_events")
    at /home/ci/build/build/sysrepo-custom/src/common.c:3011
#2  0xb6d08660 in sr_shmsub_rpc_listen_process_rpc_events (rpc_subs=0xfee7d8, conn=0x0) at /home/ci/build/build/sysrepo-custom/src/shm_sub.c:3001
#3  0xb6cda288 in sr_process_events (subscription=subscription@entry=0xfeda48, session=session@entry=0x0, stop_time_in=stop_time_in@entry=0xb66f0da8) at /home/ci/build/build/sysrepo-custom/src/sysrepo.c:2660
#4  0xb6d091b0 in sr_shmsub_listen_thread (arg=0xfeda48) at /home/ci/build/build/sysrepo-custom/src/shm_sub.c:3310
#5  0xb6e49144 in start_thread () from /lib/libpthread.so.0
#6  0xb6c2e328 in ?? () from /lib/libc.so.6
```
Git versions used:
- CESNET/libyang@1e36a94e6b9dc8077898d0aa2cefd22a6ababa78
- sysrepo@ 0e769afba31da09be6062f390335fce56f5055a9
- CESNET/libnetconf2@0722eb1f43e84eb9152ac0d5fbcfab6293fce573
- CESNET/netopeer2@2cea49605db389bd7b3805449e5329a67f826fea

Found by @peckato1

**Comment by jktjkt** [comment]
Still crashing with the latest, btw:
- CESNET/libyang@6099310f555093e5be7262df6a7f6afda76ba075
- sysrepo@ eb3615d781b1e51e0a2bd37901fbcc354f8cee71
- CESNET/netopeer2@613c7a4136311e6c7191e1db9bde87a4ef4d7afc

**Comment by michalvasko** [comment]
What platform does this occur on? The backtrace seems quite suspicious and the whole problem may simply be that the lock address was moved, now that the mutexes are robust. It would exactly fit based on the previous code. So I should be able to reproduce this quite easily, the problem with all these bugs is that the SHM does not actually change address until the segment must be reallocated, which happens only after a new memory page is required (4096B on most systems).

**Comment by jktjkt** [comment]
Thanks for looking into this. It's a 32bit ARM (Solidrun Clarfog Base, a Cortex-A9 SoC). If you have a patch which tracks and logs these addresses, I'll be happy to spin a build and share the logs. Alternatively, I can given you SSH access to the lab if you prefer so (but note that we build somewhere else than on the devices themselves, so a compile cycle will require some help from us).

**Comment by jktjkt** [comment]
Do you need more info from me, @michalvasko ? This is a blocker for us currently.

**Comment by michalvasko** [comment]
Well, you can try using this [patch](https://github.com/sysrepo/sysrepo/files/5942259/sysrepo.patch.txt) and then post the output (`netopeer2-server -d -v2`). What Unix are you running on that platform? What is the page size? I did not manage to reproduce it even when I forced the SHM to change address after remapping but I think this can easily depend on the system, embedded ones are more likely to have problems.

**Comment by jktjkt** [comment]
It's a pretty standard 32bit ARM running embedded Linux, kernel version 5.9.3, glibc userland. The page size is 4kB, and I can see that the address did change.

Here's the log, and note that I can continue after the initial `SIGILL`, so I think that's some ARMish way of checking for a supported crypto accelerator:
```gdb
# gdb --ex run --args netopeer2-server -d -v2
GNU gdb (GDB) 8.2.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "arm-buildroot-linux-gnueabihf".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from netopeer2-server...done.
Starting program: /usr/bin/netopeer2-server -d -v2
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/libthread_db.so.1".

Program received signal SIGILL, Illegal instruction.
0xb6905288 in ?? () from /lib/libcrypto.so.1.1
(gdb) bt
#0  0xb6905288 in ?? () from /lib/libcrypto.so.1.1
#1  0xb68ff1f0 in ?? () from /lib/libcrypto.so.1.1
#2  0xb6f08b2c in call_init.part () from /lib/ld-linux-armhf.so.3
#3  0xb6f08c30 in _dl_init () from /lib/ld-linux-armhf.so.3
#4  0xb6ef8a44 in _dl_start_user () from /lib/ld-linux-armhf.so.3
Backtrace stopped: previous frame identical to this frame (corrupt stack?)
(gdb) c
Continuing.
[INF]: LY: Plugin "/usr/lib/libyang1/extensions/nacm.so" successfully loaded.
[INF]: LY: Plugin "/usr/lib/libyang1/extensions/yangdata.so" successfully loaded.
[INF]: LY: Plugin "/usr/lib/libyang1/extensions/metadata.so" successfully loaded.
[INF]: LY: Plugin "/usr/lib/libyang1/user_types/user_yang_types.so" successfully loaded.
[INF]: LY: Plugin "/usr/lib/libyang1/user_types/user_inet_types.so" successfully loaded.
[INF]: LY: Reading module "ietf-yang-metadata".
[INF]: LY: Module "ietf-yang-metadata@2016-08-05" successfully parsed as implemented.
[INF]: LY: Reading module "yang".
[INF]: LY: Resolving "yang" unresolved schema nodes and their constraints...
[INF]: LY: All "yang" schema nodes and constraints resolved.
[INF]: LY: Module "yang@2017-02-20" successfully parsed as implemented.
[INF]: LY: Reading module "ietf-inet-types".
[INF]: LY: Resolving derived type "union" failed, it will be attempted later.
[INF]: LY: Resolving derived type "union" failed, it will be attempted later.
[INF]: LY: Resolving derived type "union" failed, it will be attempted later.
[INF]: LY: Resolving derived type "union" failed, it will be attempted later.
[INF]: LY: Resolving "ietf-inet-types" unresolved schema nodes and their constraints...
[INF]: LY: All "ietf-inet-types" schema nodes and constraints resolved.
[INF]: LY: Module "ietf-inet-types@2013-07-15" successfully parsed as implemented.
[INF]: LY: Reading module "ietf-yang-types".
[INF]: LY: Module "ietf-yang-types@2013-07-15" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-yang-types" in /etc/sysrepo/yang.
[INF]: LY: Resolving "sysrepo" unresolved schema nodes and their constraints...
[INF]: LY: All "sysrepo" schema nodes and constraints resolved.
[INF]: LY: Module "sysrepo@2021-01-18" successfully parsed as implemented.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[WRN]: SR: Connection with CID 2 is dead.
[INF]: SR: Applying scheduled changes.
[INF]: LY: Searching for "ietf-datastores" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-datastores@2018-02-14.yang" file.
[INF]: LY: Module "ietf-datastores@2018-02-14" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-yang-library" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-yang-library@2019-01-04.yang" file.
[INF]: LY: Searching for "ietf-inet-types" in /etc/sysrepo/yang.
[INF]: LY: Searching for "ietf-datastores" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-datastores@2018-02-14.yang" file.
[INF]: LY: Module "ietf-yang-library@2019-01-04" successfully parsed as implemented.
[INF]: LY: Searching for "sysrepo-monitoring" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/sysrepo-monitoring@2021-01-15.yang" file.
[INF]: LY: Resolving "sysrepo-monitoring" unresolved schema nodes and their constraints...
[INF]: LY: All "sysrepo-monitoring" schema nodes and constraints resolved.
[INF]: LY: Module "sysrepo-monitoring@2021-01-15" successfully parsed as implemented.
[INF]: LY: Searching for "sysrepo-plugind" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/sysrepo-plugind@2020-12-10.yang" file.
[INF]: LY: Module "sysrepo-plugind@2020-12-10" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-netconf-with-defaults" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-netconf-with-defaults@2011-06-01.yang" file.
[INF]: LY: Searching for "ietf-netconf" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-netconf@2013-09-29.yang" file.
[INF]: LY: Searching for "ietf-netconf-acm" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-netconf-acm@2018-02-14.yang" file.
[INF]: LY: Module "ietf-netconf-acm@2018-02-14" successfully parsed as imported.
[INF]: LY: Module "ietf-netconf@2013-09-29" successfully parsed as imported.
[INF]: LY: Resolving "ietf-netconf-with-defaults" unresolved schema nodes and their constraints...
[INF]: LY: All "ietf-netconf-with-defaults" schema nodes and constraints resolved.
[INF]: LY: Module "ietf-netconf-with-defaults@2011-06-01" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-netconf-notifications" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-netconf-notifications@2012-02-06.yang" file.
[INF]: LY: Resolving "ietf-netconf-notifications" unresolved schema nodes and their constraints...
[INF]: LY: All "ietf-netconf-notifications" schema nodes and constraints resolved.
[INF]: LY: Module "ietf-netconf-notifications@2012-02-06" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-origin" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-origin@2018-02-14.yang" file.
[INF]: LY: Searching for "ietf-yang-metadata" in /etc/sysrepo/yang.
[INF]: LY: Resolving "ietf-origin" unresolved schema nodes and their constraints...
[INF]: LY: All "ietf-origin" schema nodes and constraints resolved.
[INF]: LY: Module "ietf-origin@2018-02-14" successfully parsed as implemented.
[INF]: LY: Module "ietf-netconf-acm@2018-02-14" now implemented.
[INF]: LY: Searching for "ietf-netconf-monitoring" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-netconf-monitoring@2010-10-04.yang" file.
[INF]: LY: Module "ietf-netconf-monitoring@2010-10-04" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-netconf-nmda" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-netconf-nmda@2019-01-07.yang" file.
[INF]: LY: Searching for "ietf-origin" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-origin@2018-02-14.yang" file.
[INF]: LY: Searching for "ietf-netconf-with-defaults" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-netconf-with-defaults@2011-06-01.yang" file.
[INF]: LY: Resolving "ietf-netconf-nmda" unresolved schema nodes and their constraints...
[INF]: LY: All "ietf-netconf-nmda" schema nodes and constraints resolved.
[INF]: LY: Module "ietf-netconf-nmda@2019-01-07" successfully parsed as implemented.
[INF]: LY: Searching for "nc-notifications" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/nc-notifications@2008-07-14.yang" file.
[INF]: LY: Searching for "notifications" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/notifications@2008-07-14.yang" file.
[INF]: LY: Module "notifications@2008-07-14" successfully parsed as imported.
[INF]: LY: Module "nc-notifications@2008-07-14" successfully parsed as implemented.
[INF]: LY: Module "notifications@2008-07-14" now implemented.
[INF]: LY: Searching for "ietf-x509-cert-to-name" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-x509-cert-to-name@2014-12-10.yang" file.
[INF]: LY: Module "ietf-x509-cert-to-name@2014-12-10" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-crypto-types" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-crypto-types@2019-07-02.yang" file.
[INF]: LY: Module "ietf-crypto-types@2019-07-02" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-keystore" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-keystore@2019-07-02.yang" file.
[INF]: LY: Searching for "ietf-crypto-types" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-crypto-types@2019-07-02.yang" file.
[INF]: LY: Resolving "ietf-keystore" unresolved schema nodes and their constraints...
[INF]: LY: All "ietf-keystore" schema nodes and constraints resolved.
[INF]: LY: Module "ietf-keystore@2019-07-02" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-truststore" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-truststore@2019-07-02.yang" file.
[INF]: LY: Module "ietf-truststore@2019-07-02" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-tcp-common" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-tcp-common@2019-07-02.yang" file.
[INF]: LY: Module "ietf-tcp-common@2019-07-02" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-ssh-server" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-ssh-server@2019-07-02.yang" file.
[INF]: LY: Searching for "ietf-ssh-common" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-ssh-common@2019-07-02.yang" file.
[INF]: LY: Module "ietf-ssh-common@2019-07-02" successfully parsed as imported.
[INF]: LY: Searching for "ietf-keystore" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-keystore@2019-07-02.yang" file.
[INF]: LY: Searching for "iana-crypt-hash" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/iana-crypt-hash@2014-08-06.yang" file.
[INF]: LY: Module "iana-crypt-hash@2014-08-06" successfully parsed as imported.
[INF]: LY: Module "ietf-ssh-server@2019-07-02" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-tls-server" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-tls-server@2019-07-02.yang" file.
[INF]: LY: Searching for "ietf-tls-common" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-tls-common@2019-07-02.yang" file.
[INF]: LY: Module "ietf-tls-common@2019-07-02" successfully parsed as imported.
[INF]: LY: Searching for "ietf-truststore" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-truststore@2019-07-02.yang" file.
[INF]: LY: Module "ietf-tls-server@2019-07-02" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-netconf-server" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-netconf-server@2019-07-02.yang" file.
[INF]: LY: Searching for "ietf-x509-cert-to-name" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-x509-cert-to-name@2014-12-10.yang" file.
[INF]: LY: Searching for "ietf-tcp-client" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-tcp-client@2019-07-02.yang" file.
[INF]: LY: Searching for "ietf-tcp-common" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-tcp-common@2019-07-02.yang" file.
[INF]: LY: Module "ietf-tcp-client@2019-07-02" successfully parsed as imported.
[INF]: LY: Searching for "ietf-tcp-server" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-tcp-server@2019-07-02.yang" file.
[INF]: LY: Module "ietf-tcp-server@2019-07-02" successfully parsed as imported.
[INF]: LY: Resolving "ietf-netconf-server" unresolved schema nodes and their constraints...
[INF]: LY: All "ietf-netconf-server" schema nodes and constraints resolved.
[INF]: LY: Module "ietf-netconf-server@2019-07-02" successfully parsed as implemented.
[INF]: LY: Searching for "iana-hardware" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/iana-hardware@2018-03-13.yang" file.
[INF]: LY: Module "iana-hardware@2018-03-13" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-hardware" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-hardware@2018-03-13.yang" file.
[INF]: LY: Searching for "iana-hardware" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/iana-hardware@2018-03-13.yang" file.
[INF]: LY: Resolving "ietf-hardware" unresolved schema nodes and their constraints...
[INF]: LY: All "ietf-hardware" schema nodes and constraints resolved.
[INF]: LY: Module "ietf-hardware@2018-03-13" successfully parsed as implemented.
[INF]: LY: Searching for "czechlight-roadm-device" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/czechlight-roadm-device@2019-09-30.yang" file.
[INF]: LY: Searching for "opendevice-types" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/opendevice-types@2017-10-08.yang" file.
[INF]: LY: Module "opendevice-types@2017-10-08" successfully parsed as imported.
[INF]: LY: Searching for "czechlight-roadm-common" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/czechlight-roadm-common@2019-09-30.yang" file.
[INF]: LY: Resolving "czechlight-roadm-common" unresolved schema nodes and their constraints...
[INF]: LY: All "czechlight-roadm-common" schema nodes and constraints resolved.
[INF]: LY: Module "czechlight-roadm-common@2019-09-30" successfully parsed as imported.
[INF]: LY: Resolving "czechlight-roadm-device" unresolved schema nodes and their constraints...
[INF]: LY: All "czechlight-roadm-device" schema nodes and constraints resolved.
[INF]: LY: Module "czechlight-roadm-device@2019-09-30" successfully parsed as implemented.
[INF]: LY: Searching for "ietf-system" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/ietf-system@2014-08-06.yang" file.
[INF]: LY: Resolving "ietf-system" unresolved schema nodes and their constraints...
[INF]: LY: All "ietf-system" schema nodes and constraints resolved.
[INF]: LY: Module "ietf-system@2014-08-06" successfully parsed as implemented.
[INF]: LY: Searching for "czechlight-lldp" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/czechlight-lldp@2020-11-04.yang" file.
[INF]: LY: Searching for "iana-afn-safi" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/iana-afn-safi@2013-07-04.yang" file.
[INF]: LY: Module "iana-afn-safi@2013-07-04" successfully parsed as imported.
[INF]: LY: Module "czechlight-lldp@2020-11-04" successfully parsed as implemented.
[INF]: LY: Searching for "czechlight-system" in /etc/sysrepo/yang.
[INF]: LY: Loading schema from "/etc/sysrepo/yang/czechlight-system@2021-01-13.yang" file.
[INF]: LY: Module "czechlight-system@2021-01-13" successfully parsed as implemented.
[INF]: SR: No scheduled changes.
[INF]: SR: Connection 3 created.
[INF]: SR: Session 8 (user "root", CID 3) created.
[New Thread 0xb66c7450 (LWP 589)]
[WRN]: SR: Recovering RPC/action "/ietf-netconf:get-config" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/ietf-netconf:edit-config" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/ietf-netconf:copy-config" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/ietf-netconf:delete-config" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/ietf-netconf:lock" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/ietf-netconf:unlock" subscription of CID 2.
[WRN]: SR: Recovered a read-lock of CID 2 (sr_shmext_rpc_subscription_add).
[WRN]: SR: Recovering RPC/action "/ietf-netconf:get" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/ietf-netconf:kill-session" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/ietf-netconf:commit" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/ietf-netconf:discard-changes" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/ietf-netconf:validate" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/notifications:create-subscription" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/ietf-netconf-nmda:get-data" subscription of CID 2.
[WRN]: SR: Recovering RPC/action "/ietf-netconf-nmda:edit-data" subscription of CID 2.
[New Thread 0xb5ebe450 (LWP 590)]
[WRN]: SR: Recovering module "ietf-netconf-monitoring" operational subscription of CID 2.
[WRN]: SR: Recovering module "nc-notifications" operational subscription of CID 2.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: There is no leafref "/ietf-keystore:keystore/ietf-keystore:asymmetric-keys/ietf-keystore:asymmetric-key/ietf-keystore:name" with the value "genkey", but it is not required.
[INF]: LY: All data nodes and constraints resolved.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: LN: Listening on :::830 for SSH connections.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[WRN]: SR: Recovering module "ietf-netconf-server" operational subscription of CID 2.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-server" "done" event on enabled data.
[INF]: SR: Triggering "ietf-keystore" "done" event on enabled data.
[INF]: SR: Triggering "ietf-truststore" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-acm" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-acm" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-acm" "done" event on enabled data.
[INF]: SR: Triggering "ietf-netconf-acm" "done" event on enabled data.
[WRN]: SR: Recovering module "ietf-netconf-acm" operational subscription of CID 2.
[WRN]: SR: Recovering module "ietf-netconf-acm" operational subscription of CID 2.
[WRN]: SR: Recovering module "ietf-netconf-acm" operational subscription of CID 2.
[New Thread 0xb56b2450 (LWP 591)]
[New Thread 0xb4eb1450 (LWP 592)]
[New Thread 0xb46b0450 (LWP 593)]
[New Thread 0xb3cff450 (LWP 594)]
[INF]: LN: Accepted a connection on :::830.
[INF]: SR: Session 9 (user "root", CID 3) created.
[INF]: LN: Received an SSH message "request-service" of subtype "ssh-userauth".
[INF]: LN: Received an SSH message "request-auth" of subtype "none".
[INF]: LN: Received an SSH message "request-auth" of subtype "publickey".
[INF]: LN: Received an SSH message "request-auth" of subtype "publickey".
[INF]: LN: User "root" authenticated.
[INF]: LN: Received an SSH message "request-channel-open" of subtype "session".
[INF]: LN: Received an SSH message "request-channel" of subtype "subsystem".
[INF]: SR: Session 10 (user "root", CID 3) created.
[INF]: SR: There are no subscribers for "ietf-netconf-notifications" notifications.
[INF]: NP: Generated new event (netconf-session-start).
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: LN: Schema "ietf-yang-library" was requested.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: NP: Session 1: thread 0 event new RPC.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: LN: Schema "ietf-datastores" was requested.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: NP: Session 1: thread 0 event new RPC.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: LN: Schema "ietf-datastores" was requested.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: NP: Session 1: thread 2 event new RPC.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: SR: Published event "rpc" "/ietf-netconf:get" with ID 1 priority 0 for 1 subscribers.
[INF]: SR: Processing "/ietf-netconf:get" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[INF]: LY: Resolving unresolved data nodes and their constraints...
[INF]: LY: All data nodes and constraints resolved.
[WRN]: SR: ## locked on addr 0xb5ec6000
[INF]: SR: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: SR: ## unlocking addr 0xb3eab000

Thread 2 "netopeer2-serve" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xb66c7450 (LWP 589)]
0xb6e244c0 in __pthread_mutex_unlock_full () from /lib/libpthread.so.0
(gdb) t a a bt

Thread 7 (Thread 0xb3cff450 (LWP 594)):
#0  0xb6e24ee4 in __pthread_mutex_cond_lock_full () from /lib/libpthread.so.0
#1  0xb6e28278 in pthread_cond_timedwait@@GLIBC_2.4 () from /lib/libpthread.so.0
#2  0xb6cdad5c in sr_shmsub_notify_wait_wr (sub_shm=sub_shm@entry=0xb3eaf000, shm_struct_size=shm_struct_size@entry=176, expected_ev=expected_ev@entry=SR_SUB_EV_ERROR, timeout_ms=timeout_ms@entry=1, cid=3, cb_err_info=0xb3cfec14, 
    cb_err_info@entry=0xb66f5a88) at /home/jkt/work/prog/_build/_br-cfb/build/sysrepo-custom/src/shm_sub.c:248
#3  0xb6cddc00 in sr_shmsub_rpc_notify (conn=0x32ea58, shm_rpc=0xb66f5a60, shm_rpc@entry=0xb3cfec14, op_path=0x3e8fb8 "/ietf-netconf:get", input=input@entry=0x34ce08, sid=..., timeout_ms=timeout_ms@entry=2000, request_id=0xb3cfec24, 
    request_id@entry=0xb3cfec1c, output=output@entry=0xb3cfec90, cb_err_info=0xb3cfec14, cb_err_info@entry=0xb3cfec0c) at /home/jkt/work/prog/_build/_br-cfb/build/sysrepo-custom/src/shm_sub.c:1620
#4  0xb6cb264c in sr_rpc_send_tree (session=0xb3d16218, input=0x34ce08, timeout_ms=2000, output=0xb3cfec90) at /home/jkt/work/prog/_build/_br-cfb/build/sysrepo-custom/src/sysrepo.c:3883
#5  0x00015c78 in np2srv_rpc_cb (rpc=0x34ce08, ncs=0xb3d007e0) at /home/jkt/work/prog/_build/_br-cfb/build/netopeer2-custom/src/main.c:444
#6  0xb6ec5288 in nc_server_send_reply_io (rpc=0x3949e0, io_timeout=10, session=0xb3d007e0) at /home/jkt/work/prog/_build/_br-cfb/build/libnetconf2-custom/src/session_server.c:1305
#7  nc_ps_poll (ps=<optimized out>, timeout=10, session=0xb6eec000) at /home/jkt/work/prog/_build/_br-cfb/build/libnetconf2-custom/src/session_server.c:1656
#8  0x00018164 in worker_thread (arg=0x34c088) at /home/jkt/work/prog/_build/_br-cfb/build/netopeer2-custom/src/main.c:994
#9  0xb6e20144 in start_thread () from /lib/libpthread.so.0
#10 0xb6c05328 in ?? () from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 6 (Thread 0xb46b0450 (LWP 593)):
#0  0xb6bcacb4 in ?? () from /lib/libc.so.6
#1  0xb6bcacfc in clock_nanosleep () from /lib/libc.so.6
#2  0xb6bd0b50 in nanosleep () from /lib/libc.so.6
#3  0x00019464 in np_sleep (ms=200) at /home/jkt/work/prog/_build/_br-cfb/build/netopeer2-custom/src/common.c:50
#4  0x00018190 in worker_thread (arg=0x391920) at /home/jkt/work/prog/_build/_br-cfb/build/netopeer2-custom/src/main.c:998
#5  0xb6e20144 in start_thread () from /lib/libpthread.so.0
#6  0xb6c05328 in ?? () from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 5 (Thread 0xb4eb1450 (LWP 592)):
#0  0xb6bcacb4 in ?? () from /lib/libc.so.6
#1  0xb6bcacfc in clock_nanosleep () from /lib/libc.so.6
#2  0xb6bd0b50 in nanosleep () from /lib/libc.so.6
#3  0x00019464 in np_sleep (ms=200) at /home/jkt/work/prog/_build/_br-cfb/build/netopeer2-custom/src/common.c:50
#4  0x00018190 in worker_thread (arg=0x38a328) at /home/jkt/work/prog/_build/_br-cfb/build/netopeer2-custom/src/main.c:998
#5  0xb6e20144 in start_thread () from /lib/libpthread.so.0
#6  0xb6c05328 in ?? () from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 4 (Thread 0xb56b2450 (LWP 591)):
#0  0xb6bcacb8 in ?? () from /lib/libc.so.6
#1  0xb6bcacfc in clock_nanosleep () from /lib/libc.so.6
#2  0xb6bd0b50 in nanosleep () from /lib/libc.so.6
#3  0xb6bff630 in usleep () from /lib/libc.so.6
#4  0xb6ec49c8 in nc_ps_poll (ps=0x349b38, timeout=10, session=0xb6eec000) at /home/jkt/work/prog/_build/_br-cfb/build/libnetconf2-custom/src/session_server.c:1610
#5  0x00018164 in worker_thread (arg=0x397b30) at /home/jkt/work/prog/_build/_br-cfb/build/netopeer2-custom/src/main.c:994
#6  0xb6e20144 in start_thread () from /lib/libpthread.so.0
#7  0xb6c05328 in ?? () from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 3 (Thread 0xb5ebe450 (LWP 590)):
#0  0xb6bfee90 in select () from /lib/libc.so.6
#1  0xb6ce06c4 in sr_shmsub_listen_thread (arg=0x34ca28) at /home/jkt/work/prog/_build/_br-cfb/build/sysrepo-custom/src/shm_sub.c:3337
#2  0xb6e20144 in start_thread () from /lib/libpthread.so.0
#3  0xb6c05328 in ?? () from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 2 (Thread 0xb66c7450 (LWP 589)):
#0  0xb6e244c0 in __pthread_mutex_unlock_full () from /lib/libpthread.so.0
#1  0xb6cb95e0 in sr_rwunlock (rwlock=0xb3eab000, rwlock@entry=0x391b64, timeout_ms=timeout_ms@entry=0, mode=mode@entry=SR_LOCK_WRITE, cid=3, func=0xb6ce6e9c <__func__.3> "sr_shmsub_rpc_listen_process_rpc_events")
    at /home/jkt/work/prog/_build/_br-cfb/build/sysrepo-custom/src/common.c:3241
#2  0xb6cdfbac in sr_shmsub_rpc_listen_process_rpc_events (rpc_subs=0x391b58, conn=0x0) at /home/jkt/work/prog/_build/_br-cfb/build/sysrepo-custom/src/shm_sub.c:3005
#3  0xb6cb128c in sr_process_events (subscription=subscription@entry=0x34a358, session=session@entry=0x0, stop_time_in=stop_time_in@entry=0xb66c6da8) at /home/jkt/work/prog/_build/_br-cfb/build/sysrepo-custom/src/sysrepo.c:2661
#4  0xb6ce0700 in sr_shmsub_listen_thread (arg=0x34a358) at /home/jkt/work/prog/_build/_br-cfb/build/sysrepo-custom/src/shm_sub.c:3314
#5  0xb6e20144 in start_thread () from /lib/libpthread.so.0
#6  0xb6c05328 in ?? () from /lib/libc.so.6
Backtrace stopped: previous frame identical to this frame (corrupt stack?)

Thread 1 (Thread 0xb6f23220 (LWP 586)):
#0  0xb6bcacb4 in ?? () from /lib/libc.so.6
#1  0xb6bcacfc in clock_nanosleep () from /lib/libc.so.6
#2  0xb6bd0b50 in nanosleep () from /lib/libc.so.6
#3  0x00019464 in np_sleep (ms=200) at /home/jkt/work/prog/_build/_br-cfb/build/netopeer2-custom/src/common.c:50
#4  0x00018190 in worker_thread (arg=0x3ea9b0) at /home/jkt/work/prog/_build/_br-cfb/build/netopeer2-custom/src/main.c:998
#5  0x00019254 in main (argc=3, argv=0xbecabd94) at /home/jkt/work/prog/_build/_br-cfb/build/netopeer2-custom/src/main.c:1337
```

**Comment by michalvasko** [comment]
Okay, can you try running *sysrepo* `test_rpc_action`? If it is really the problem of changing the address, it should crash, too. Or it should be seen that the addresses do not change (for one reason or another) and then it will not crash. In any case, it should help determine the exact problem.

**Comment by jktjkt** [comment]
We never run tests during the cross build itself, but I'll see if I can build the test binary and run it on the target HW.

**Comment by jktjkt** [comment]
Is there a script (that is not CMake) which installs all the required YANG modules?

**Comment by michalvasko** [comment]
No, the modules are installed by the test itself.

**Comment by jktjkt** [comment]
OK, I created `/home/jkt/work/prog/_build/_br-cfb/build/sysrepo-custom/tests/files` on the target HW with a copy of the YANG files from the test dir. It seems that the test passes:

```console-session
# LD_LIBRARY_PATH=/tmp /tmp/test_rpc_action 
[==========] Running 12 test(s).
[INF]: Applying scheduled changes.
[INF]: No scheduled changes.
[INF]: Connection 18 created.
[INF]: Module "test" scheduled for installation.
[INF]: Module "ietf-interfaces" scheduled for installation.
[INF]: Module "iana-if-type" scheduled for installation.
[INF]: File "ietf-interfaces@2014-05-08.yang" was installed.
[INF]: Module "ops-ref" scheduled for installation.
[INF]: Module "ops" scheduled for installation.
[INF]: File "ops-ref.yang" was installed.
[INF]: Module "act" scheduled for installation.
[INF]: Module "act2" scheduled for installation.
[INF]: File "act@2019-04-30.yang" was installed.
[INF]: Module "act3" scheduled for installation.
[INF]: File "act2@2019-04-30.yang" was installed.
[INF]: Applying scheduled changes.
[INF]: File "test.yang" was installed.
[INF]: Module "test" was installed.
[INF]: Module "ietf-interfaces" will be installed as "iana-if-type" module dependency.
[INF]: File "iana-if-type@2014-05-08.yang" was installed.
[INF]: Module "iana-if-type" was installed.
[INF]: Dependency module "ietf-interfaces" was installed.
[INF]: Module "ops-ref" will be installed as "ops" module dependency.
[INF]: File "ops.yang" was installed.
[INF]: Module "ops" was installed.
[INF]: Dependency module "ops-ref" was installed.
[INF]: Module "act" will be installed as "act2" module dependency.
[INF]: Module "act2" will be installed as "act3" module dependency.
[INF]: File "act3@2019-04-30.yang" was installed.
[INF]: Module "act3" was installed.
[INF]: Dependency module "act" was installed.
[INF]: Dependency module "act2" was installed.
[INF]: Scheduled changes applied.
[INF]: Connection 19 created.
[INF]: Session 13 (user "root", CID 19) created.
[ RUN      ] test_fail
[INF]: Published event "rpc" "/ops:rpc1" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:rpc1" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Failed processing of "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[ERR]: RPC FAIL
[WRN]: Event "rpc" with ID 1 priority 0 failed (System function call failed).
[ERR]: Nested operation "act1" data parent does not exist in the operational datastore. (path: /ops:cont/list1[k='1']/cont2)
[       OK ] test_fail
[ RUN      ] test_rpc
[INF]: There are no subscribers for changes of the module "ops-ref" in running DS.
[ERR]: Leafref "/ops-ref:l1" of value "l1-val" points to a non-existing leaf. (path: /ops:rpc1/l1)
[ERR]: RPC input validation failed.
[INF]: Published event "rpc" "/ops:rpc1" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:rpc1" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[INF]: Published event "rpc" "/ops:rpc2" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:rpc2" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6efe000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6efe000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[ERR]: Leafref "/ops-ref:l2" of value "inval-ref" points to a non-existing leaf. (path: /ops:rpc2/cont/l3)
[ERR]: RPC output validation failed.
[INF]: Published event "rpc" "/ops:rpc2" with ID 2 priority 0 for 1 subscribers.
[INF]: Processing "/ops:rpc2" "rpc" event with ID 2 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6efe000
[INF]: Successful processing of "rpc" event with ID 2 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6efe000
[INF]: Event "rpc" with ID 2 priority 0 succeeded.
[INF]: Published event "rpc" "/ops:rpc3" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6efd000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6efd000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[ERR]: Schema node not found. (path: /ops:invalid)
[INF]: There are no subscribers for changes of the module "ops-ref" in running DS.
[       OK ] test_rpc
[ RUN      ] test_action
[INF]: There are no subscribers for changes of the module "ops" in running DS.
[INF]: Published event "rpc" "/ops:cont/list1/cont2/act1" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:cont/list1/cont2/act1" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[INF]: Published event "rpc" "/ops:cont/list1/act2" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:cont/list1/act2" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6efe000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6efe000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[INF]: There are no subscribers for changes of the module "ops" in running DS.
[       OK ] test_action
[ RUN      ] test_action_pred
[ERR]: Unexpected character(s) ' ' (" or /ops:rpc1").
[INF]: There are no subscribers for changes of the module "ops" in running DS.
[ERR]: There are no matching subscribers for RPC/action "/ops:cont/list1/cont2/act1". (path: /ops:cont/list1/cont2/act1)
[INF]: Published event "rpc" "/ops:cont/list1/cont2/act1" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:cont/list1/cont2/act1" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[INF]: Published event "rpc" "/ops:cont/list1/cont2/act1" with ID 2 priority 1 for 1 subscribers.
[INF]: Processing "/ops:cont/list1/cont2/act1" "rpc" event with ID 2 priority 1 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 2 priority 1 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 2 priority 1 succeeded.
[INF]: There are no subscribers for changes of the module "ops" in running DS.
[       OK ] test_action_pred
[ RUN      ] test_multi
[INF]: There are no subscribers for changes of the module "ops" in running DS.
[INF]: Published event "rpc" "/ops:cont/list1/cont2/act1" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:cont/list1/cont2/act1" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[INF]: Published event "rpc" "/ops:cont/list1/cont2/act1" with ID 2 priority 1 for 1 subscribers.
[INF]: Processing "/ops:cont/list1/cont2/act1" "rpc" event with ID 2 priority 1 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 2 priority 1 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 2 priority 1 succeeded.
[INF]: Published event "rpc" "/ops:cont/list1/cont2/act1" with ID 2 priority 0 for 1 subscribers.
[INF]: Processing "/ops:cont/list1/cont2/act1" "rpc" event with ID 2 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 2 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 2 priority 0 succeeded.
[INF]: Published event "rpc" "/ops:cont/list1/cont2/act1" with ID 3 priority 2 for 1 subscribers.
[INF]: Processing "/ops:cont/list1/cont2/act1" "rpc" event with ID 3 priority 2 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 3 priority 2 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 3 priority 2 succeeded.
[INF]: Published event "rpc" "/ops:cont/list1/cont2/act1" with ID 3 priority 1 for 1 subscribers.
[INF]: Processing "/ops:cont/list1/cont2/act1" "rpc" event with ID 3 priority 1 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 3 priority 1 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 3 priority 1 succeeded.
[INF]: Published event "rpc" "/ops:cont/list1/cont2/act1" with ID 3 priority 0 for 1 subscribers.
[INF]: Processing "/ops:cont/list1/cont2/act1" "rpc" event with ID 3 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 3 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 3 priority 0 succeeded.
[INF]: There are no subscribers for changes of the module "ops" in running DS.
[       OK ] test_multi
[ RUN      ] test_multi_fail
[INF]: Published event "rpc" "/ops:rpc3" with ID 1 priority 2 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 1 priority 2 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 1 priority 2 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 1 priority 2 succeeded.
[INF]: Published event "rpc" "/ops:rpc3" with ID 1 priority 1 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 1 priority 1 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 1 priority 1 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 1 priority 1 succeeded.
[INF]: Published event "rpc" "/ops:rpc3" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Failed processing of "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[ERR]: Out of memory
[WRN]: Event "rpc" with ID 1 priority 0 failed (Out of memory).
[INF]: Published event "abort" "/ops:rpc3" with ID 1 priority 2 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "abort" event with ID 1 priority 2 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "abort" event with ID 1 priority 2 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Published event "abort" "/ops:rpc3" with ID 1 priority 1 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "abort" event with ID 1 priority 1 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "abort" event with ID 1 priority 1 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Published event "rpc" "/ops:rpc3" with ID 2 priority 2 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 2 priority 2 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 2 priority 2 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 2 priority 2 succeeded.
[INF]: Published event "rpc" "/ops:rpc3" with ID 2 priority 1 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 2 priority 1 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Failed processing of "rpc" event with ID 2 priority 1 (remaining 1 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[ERR]: Item not found
[WRN]: Event "rpc" with ID 2 priority 1 failed (Item not found).
[INF]: Published event "abort" "/ops:rpc3" with ID 2 priority 2 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "abort" event with ID 2 priority 2 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "abort" event with ID 2 priority 2 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Published event "rpc" "/ops:rpc3" with ID 3 priority 2 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 3 priority 2 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Failed processing of "rpc" event with ID 3 priority 2 (remaining 1 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[ERR]: Requested resource already locked
[WRN]: Event "rpc" with ID 3 priority 2 failed (Requested resource already locked).
[INF]: Published event "rpc" "/ops:rpc3" with ID 4 priority 2 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 4 priority 2 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 4 priority 2 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 4 priority 2 succeeded.
[INF]: Published event "rpc" "/ops:rpc3" with ID 4 priority 1 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 4 priority 1 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 4 priority 1 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 4 priority 1 succeeded.
[INF]: Published event "rpc" "/ops:rpc3" with ID 4 priority 0 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 4 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 4 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 4 priority 0 succeeded.
[       OK ] test_multi_fail
[ RUN      ] test_action_deps
[ERR]: Nested operation "conditional_action" data parent does not exist in the operational datastore. (path: /act:advanced/act3:conditional)
[INF]: Published event "rpc" "/act:advanced/act3:conditional/conditional_action" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/act:advanced/act3:conditional/conditional_action" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[INF]: Published event "rpc" "/act:advanced/act3:conditional_action2" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/act:advanced/act3:conditional_action2" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6efe000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6efe000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[       OK ] test_action_deps
[ RUN      ] test_action_change_config
[INF]: There are no subscribers for changes of the module "ops" in running DS.
[INF]: Published event "rpc" "/ops:cont/list1/cont2/act1" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:cont/list1/cont2/act1" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[INF]: Published event "change" "ops" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "ops" "change" event with ID 1 priority 0 (remaining 1 subscribers).
[INF]: Successful processing of "change" event with ID 1 priority 0 (remaining 0 subscribers).
[INF]: Event "change" with ID 1 priority 0 succeeded.
[INF]: Published event "done" "ops" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "ops" "done" event with ID 1 priority 0 (remaining 1 subscribers).
[INF]: Successful processing of "done" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[INF]: There are no subscribers for changes of the module "ops" in running DS.
[       OK ] test_action_change_config
[ RUN      ] test_rpc_shelve
[INF]: Session 14 (user "root", CID 19) created.
[INF]: Published event "rpc" "/ops:rpc3" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[INF]: Shelved processing of "rpc" event with ID 1 priority 0.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[       OK ] test_rpc_shelve
[ RUN      ] test_input_parameters
[ERR]: Invalid arguments for function "_sr_rpc_subscribe".
[ERR]: Unterminated predicate in expression.
[ERR]: Unexpected character ']'(]/con) in expression.
[ERR]: Unterminated literal in expression.
[ERR]: Invalid xpath "cont/list1[k='one']/cont2/act1".
[ERR]: Invalid xpath "/cont/list1[k='one']/cont2/act1".
[ERR]: Invalid xpath ":cont/list1[k='one']/cont2/act1".
[ERR]: Invalid xpath "/1_ops:cont/list1[k='one']/cont2/act1".
[ERR]: Invalid xpath "/ops$:cont/list1[k='one']/cont2/act1".
[ERR]: Unexpected character(s) '/' ("/ops:cont/list1/cont2/act1").
[ERR]: Module "no-mod" was not found in sysrepo.
[ERR]: Schema node not found. (path: /ops:cont/list1/cont2/no-node)
[ERR]: Path "/ops:cont" does not identify an RPC nor an action.
[ERR]: Data trees must be created using the session connection libyang context.
[ERR]: Provided input is not a valid RPC or action invocation.
[ERR]: RPC subscription for "/ops:rpc1" with priority 5 already exists.
[       OK ] test_input_parameters
[ RUN      ] test_rpc_action_with_no_thread
[INF]: There are no subscribers for changes of the module "ops-ref" in running DS.
[INF]: Published event "rpc" "/ops:rpc1" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:rpc1" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6eff000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6eff000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[INF]: Published event "rpc" "/ops:rpc2" with ID 1 priority 0 for 1 subscribers.
[ERR]: Callback event "rpc" with ID 1 processing timed out.
[WRN]: Event "rpc" with ID 1 priority 0 failed (Timeout expired).
[INF]: Published event "rpc" "/ops:rpc3" with ID 1 priority 0 for 1 subscribers.
[ERR]: Callback event "rpc" with ID 1 processing timed out.
[WRN]: Event "rpc" with ID 1 priority 0 failed (Timeout expired).
[INF]: There are no subscribers for changes of the module "ops" in running DS.
[INF]: Published event "rpc" "/ops:cont/list1/cont2/act1" with ID 1 priority 0 for 1 subscribers.
[ERR]: Callback event "rpc" with ID 1 processing timed out.
[WRN]: Event "rpc" with ID 1 priority 0 failed (Timeout expired).
[       OK ] test_rpc_action_with_no_thread
[ RUN      ] test_rpc_oper
[INF]: Published event "rpc" "/ops:rpc3" with ID 1 priority 0 for 1 subscribers.
[INF]: Processing "/ops:rpc3" "rpc" event with ID 1 priority 0 (remaining 1 subscribers).
[WRN]: ## locked on addr 0xb6efe000
[INF]: Successful processing of "rpc" event with ID 1 priority 0 (remaining 0 subscribers).
[WRN]: ## unlocking addr 0xb6efe000
[INF]: Event "rpc" with ID 1 priority 0 succeeded.
[       OK ] test_rpc_oper
[INF]: Module "act3" scheduled for deletion.
[INF]: Module "act2" scheduled for deletion.
[INF]: Module "act" scheduled for deletion.
[INF]: Module "ops" scheduled for deletion.
[INF]: Module "ops-ref" scheduled for deletion.
[INF]: Module "iana-if-type" scheduled for deletion.
[INF]: Module "ietf-interfaces" scheduled for deletion.
[INF]: Module "test" scheduled for deletion.
[==========] 12 test(s) run.
[  PASSED  ] 12 test(s).
```

**Comment by michalvasko** [comment]
Okay, thanks. Yeah, it passes but it never remaps the memory. But we can assume that is the cause of the problem so I will try fix that. I hope tomorrow it could be implemented, not sure.

**Comment by michalvasko** [comment]
Please try it again now, the locks should no longer be remapped.

**Comment by jktjkt** [comment]
Thanks, the basic operations are working again -- I appreciate your help.

There's a lot of TSAN warnings, including reports of unlocking a mutex that is not unlocked. Would you like me to file them as bugs, or are you able to reproduce them as well if you try?

**Comment by michalvasko** [comment]
The mutex problem was likely solved, it was just a forgotten redundant remap. As for the rest, you can create a new issue with your output because you may have some warnings that I am not able to reproduce like the last time.

**Comment by jktjkt** [comment]
Thanks, it's been substantially improved, on a quick glance I'm now only getting data race reports via sysrepo's own testsuite. I'll take a look at these and report bug in another issue. Thanks for your support and for fixing this one!
